README.md
# Rollup plugin to serve the bundle

## Explain

Keep updates active, fix bugs at a loss, and actively incorporate PR. Based on rollup-plugin-serve version 1.0.1.

Welcome to use rollup-plugin-serve2!


<a href="LICENSE">
  <img src="https://img.shields.io/badge/license-MIT-brightgreen.svg" alt="Software License" />
</a>
<a href="https://github.com/GGICE/rollup-plugin-serve/issues">
  <img src="https://img.shields.io/github/issues/GGICE/rollup-plugin-serve.svg" alt="Issues" />
</a>
<a href="http://standardjs.com/">
  <img src="https://img.shields.io/badge/code%20style-standard-brightgreen.svg" alt="JavaScript Style Guide" />
</a>
<a href="https://npmjs.org/package/rollup-plugin-serve">
  <img src="https://img.shields.io/npm/v/rollup-plugin-serve2.svg?style=flat-squar" alt="NPM" />
</a>

## Installation
```
# Rollup v0.60+ and v1+
npm install --save-dev rollup-plugin-serve2

# Rollup v0.59 and below
npm install --save-dev rollup-plugin-serve2@0
```

## Usage
```js
// rollup.config.js
import serve from 'rollup-plugin-serve2'

export default {
  input: 'src/main.js',
  output: {
    file: 'dist/bundle.js',
    format: ...
  },
  plugins: [
    serve('dist')
  ]
}
```

### Options

By default it serves the current project folder. Change it by passing a string:
```js
serve('public')    // will be used as contentBase

// Default options
serve({
  // Launch in browser (default: false)
  open: true,

  // Page to navigate to when opening the browser.
  // Will not do anything if open=false.
  // Remember to start with a slash.
  openPage: '/different/page',

  // Show server address in console (default: true)
  verbose: false,

  // Folder to serve files from
  contentBase: '',

  // Multiple folders to serve from
  contentBase: ['dist', 'static'],

  // Set to true to return index.html (200) instead of error page (404)
  historyApiFallback: false,

  // Path to fallback page
  historyApiFallback: '/200.html',

  // Options used in setting up server
  host: 'localhost',
  port: 10001,

  // By default server will be served over HTTP (https: false). It can optionally be served over HTTPS
  https: {
    key: fs.readFileSync('/path/to/server.key'),
    cert: fs.readFileSync('/path/to/server.crt'),
    ca: fs.readFileSync('/path/to/ca.pem')
  },

  //set headers
  headers: {
    'Access-Control-Allow-Origin': '*',
    foo: 'bar'
  },

  // set custom mime types, usage https://github.com/broofa/mime#mimedefinetypemap-force--false
  mimeTypes: {
    'application/javascript': ['js_commonjs-proxy']
  }

  // execute function after server has begun listening
  onListening: function (server) {
    const address = server.address()
    const host = address.address === '::' ? 'localhost' : address.address
    // by using a bound function, we can access options as `this`
    const protocol = this.https ? 'https' : 'http'
    console.log(`Server listening at ${protocol}://${host}:${address.port}/`)
  }

  // Set up simple proxy
  // this will route all traffic starting with
  // `/api` to http://localhost:8181/api
  proxy: {
    api: 'http://localhost:8181'
  }
})
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Contributions and feedback are very welcome.

This project aims to stay lean and close to standards. New features should encourage to stick to standards. Options that match the behaviour of [webpack-dev-server](https://webpack.js.org/configuration/dev-server/#devserver) are always ok.

To get it running:
  1. Clone the project.
  2. `npm install`
  3. `npm run build`

## Credits

- [Thomas Ghysels](https://github.com/thgh)
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE) for more information.

[link-author]: https://github.com/thgh
[link-contributors]: ../../contributors
[rollup-plugin-serve2]: https://www.npmjs.com/package/rollup-plugin-serve2


CHANGELOG.md
# Changelog

All notable changes to `rollup-plugin-serve` will be documented in this file.

## [1.1.1] - 2021-10-12
### Added
- Add `proxy` support

## [1.1.0] - 2020-11-01
### Added
- Add `onListening` hook #69 @filoxo
- Show error message when port is in use #60 @jaeh

## [1.0.4] - 2020-08-28
### Added
- Add `mimeTypes` option #58 @GerardRodes

### Fixed
- Allow opening the browser even when verbose mode is disabled #64 @Richienb

## [1.0.3] - 2020-07-21
### Fixed
- Fix path.normalize error on Windows

## [1.0.2] - 2020-07-17
### Fixed
- Fix path traversal issue

## [1.0.1] - 2019-01-27
### Added
- Add Intellisense support #34

### Fixed
- Set minimum version for `mime` package

## [1.0.0] - 2019-01-11
### Fixed
- Update `ongenerate` to `generateBundle`

### Removed
- Remove built-in `favicon.ico` #20

## [0.6.1] - 2018-12-23
### Added
- Add support for `rollup --watch` (Release http server on rollup reload)

## [0.5.0] - 2018-09-18
### Added
- Allow to override path for historyApiFallback option
- Option `openPage`

### Fixed
- Fix host option

## [0.4.2] - 2017-08-25
### Added
- Option `https`

## [0.4.1] - 2017-08-16
### Added
- Option `headers`

### Fixed
- Close the server when a termination signal is encountered

## [0.4.0] - 2017-07-02
### Fixed
- Various fixes to contentBase handling

## [0.3.0] - 2017-04-07
### Changed
- Allow to pass array as contentBase

## [0.2.0] - 2017-04-07
### Changed
- Show console message only once

## [0.1.0] - 2016-09-29
### Added
- Option `open` open the project in browser
- Option `historyApiFallback`
- Default green favicon

## Changed
- Option `root` is now `contentBase`

## [0.0.1] - 2016-09-24
### Added
- Initial version

[Unreleased]: https://github.com/thgh/rollup-plugin-serve/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.6.1...v1.0.0
[0.6.1]: https://github.com/thgh/rollup-plugin-serve/compare/v0.5.0...v0.6.1
[0.5.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.4.2...v0.5.0
[0.4.2]: https://github.com/thgh/rollup-plugin-serve/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/thgh/rollup-plugin-serve/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/thgh/rollup-plugin-serve/compare/v0.0.1...v0.1.0
[0.0.1]: https://github.com/thgh/rollup-plugin-serve/releases


src/index.js
import { readFile } from 'fs'
import https, { createServer as createHttpsServer } from 'https'
import http, { createServer } from 'http'
import { resolve, posix } from 'path'

import mime from 'mime'
import opener from 'opener'

let server

/**
 * Serve your rolled up bundle like webpack-dev-server
 * @param {ServeOptions|string|string[]} options
 */
function serve (options = { contentBase: '' }) {
  if (Array.isArray(options) || typeof options === 'string') {
    options = { contentBase: options }
  }
  options.contentBase = Array.isArray(options.contentBase) ? options.contentBase : [options.contentBase || '']
  options.port = options.port || 10001
  options.headers = options.headers || {}
  options.https = options.https || false
  options.openPage = options.openPage || ''
  options.onListening = options.onListening || function noop () { }
  options.proxy = options.proxy || {}
  mime.default_type = 'text/plain'

  if (options.mimeTypes) {
    mime.define(options.mimeTypes, true)
  }

  const proxies = Object.keys(options.proxy).map(proxy => ({
    proxy,
    destination: options.proxy[proxy],
    // eslint-disable-next-line
    test: new RegExp(`\/${proxy}`)
  }))

  const requestListener = (request, response) => {
    // Remove querystring
    const unsafePath = decodeURI(request.url.split('?')[0])

    // Don't allow path traversal
    const urlPath = posix.normalize(unsafePath)

    Object.keys(options.headers).forEach((key) => {
      response.setHeader(key, options.headers[key])
    })

    // Find the appropriate proxy for the request if one exists
    const proxy = proxies.find(({ test }) => request.url.match(test))

    // If a proxy exists, forward the request to the appropriate server
    if (proxy && proxy.destination) {
      const { destination } = proxy
      const newDestination = `${destination}${request.url}`
      const { headers, method } = request

      // Get the request contents
      let body = ''
      request.on('data', chunk => { body += chunk })

      // Forward the request
      request.on('end', () => {
        const httpServer = newDestination.startsWith('https://') ? https : http
        const proxyRequest = httpServer
          .request(newDestination, {
            headers: Object.assign({}, headers, {
              host: (new URL(newDestination)).host
            }),
            method
          }, (proxyResponse) => {
            const data = []
            proxyResponse.on('data', chunk => data.push(chunk))
            proxyResponse.on('end', () => {
              Object.keys(proxyResponse.headers).forEach(key => response.setHeader(key, proxyResponse.headers[key]))
              foundProxy(response, proxyResponse.statusCode, Buffer.concat(data))
            })
          }).end(body, 'utf-8')

        proxyRequest.on('error', err => console.error(`There was a problem with the request for ${request.url}: ${err}`))
        proxyRequest.end()
      })
      return
    }

    readFileFromContentBase(options.contentBase, urlPath, function (error, content, filePath) {
      if (!error) {
        return found(response, filePath, content)
      }
      if (error.code !== 'ENOENT') {
        response.writeHead(500)
        response.end('500 Internal Server Error' +
          '\n\n' + filePath +
          '\n\n' + Object.values(error).join('\n') +
          '\n\n(rollup-plugin-serve)', 'utf-8')
        return
      }
      if (options.historyApiFallback) {
        const fallbackPath = typeof options.historyApiFallback === 'string' ? options.historyApiFallback : '/index.html'
        readFileFromContentBase(options.contentBase, fallbackPath, function (error, content, filePath) {
          if (error) {
            notFound(response, filePath)
          } else {
            found(response, filePath, content)
          }
        })
      } else {
        notFound(response, filePath)
      }
    })
  }

  // release previous server instance if rollup is reloading configuration in watch mode
  if (server) {
    server.close()
  } else {
    closeServerOnTermination()
  }

  // If HTTPS options are available, create an HTTPS server
  server = options.https
    ? createHttpsServer(options.https, requestListener)
    : createServer(requestListener)
  server.listen(options.port, options.host, () => options.onListening(server))

  // Assemble url for error and info messages
  const url = (options.https ? 'https' : 'http') + '://' + (options.host || 'localhost') + ':' + options.port

  // Handle common server errors
  server.on('error', e => {
    if (e.code === 'EADDRINUSE') {
      console.error(url + ' is in use, either stop the other server or use a different port.')
      process.exit()
    } else {
      throw e
    }
  })

  let first = true

  return {
    name: 'serve',
    generateBundle () {
      if (first) {
        first = false

        // Log which url to visit
        if (options.verbose !== false) {
          options.contentBase.forEach(base => {
            console.log(green(url) + ' -> ' + resolve(base))
          })
        }

        // Open browser
        if (options.open) {
          if (/https?:\/\/.+/.test(options.openPage)) {
            opener(options.openPage)
          } else {
            opener(url + options.openPage)
          }
        }
      }
    }
  }
}

function readFileFromContentBase (contentBase, urlPath, callback) {
  let filePath = resolve(contentBase[0] || '.', '.' + urlPath)

  // Load index.html in directories
  if (urlPath.endsWith('/')) {
    filePath = resolve(filePath, 'index.html')
  }

  readFile(filePath, (error, content) => {
    if (error && contentBase.length > 1) {
      // Try to read from next contentBase
      readFileFromContentBase(contentBase.slice(1), urlPath, callback)
    } else {
      // We know enough
      callback(error, content, filePath)
    }
  })
}

function notFound (response, filePath) {
  response.writeHead(404)
  response.end('404 Not Found' +
    '\n\n' + filePath +
    '\n\n(rollup-plugin-serve)', 'utf-8')
}

function found (response, filePath, content) {
  response.writeHead(200, { 'Content-Type': mime.getType(filePath) })
  response.end(content, 'utf-8')
}

function green (text) {
  return '\u001b[1m\u001b[32m' + text + '\u001b[39m\u001b[22m'
}

function foundProxy (response, status, content) {
  response.writeHead(status)
  response.end(content, 'utf-8')
}

function closeServerOnTermination () {
  const terminationSignals = ['SIGINT', 'SIGTERM', 'SIGQUIT', 'SIGHUP']
  terminationSignals.forEach(signal => {
    process.on(signal, () => {
      if (server) {
        server.close()
        process.exit()
      }
    })
  })
}

export default serve

/**
 * @typedef {Object} ServeOptions
 * @property {boolean} [open=false] Launch in browser (default: `false`)
 * @property {string} [openPage=''] Page to navigate to when opening the browser. Will not do anything if `open` is `false`. Remember to start with a slash e.g. `'/different/page'`
 * @property {boolean} [verbose=true] Show server address in console (default: `true`)
 * @property {string|string[]} [contentBase=''] Folder(s) to serve files from
 * @property {string|boolean} [historyApiFallback] Path to fallback page. Set to `true` to return index.html (200) instead of error page (404)
 * @property {string} [host='localhost'] Server host (default: `'localhost'`)
 * @property {number} [port=10001] Server port (default: `10001`)
 * @property {function} [onListening] Execute a function when server starts listening for connections on a port
 * @property {ServeOptionsHttps} [https=false] By default server will be served over HTTP (https: `false`). It can optionally be served over HTTPS
 * @property {{[header:string]: string}} [headers] Set headers
 */

/**
 * @typedef {Object} ServeOptionsHttps
 * @property {string|Buffer|Buffer[]|Object[]} key
 * @property {string|Buffer|Array<string|Buffer>} cert
 * @property {string|Buffer|Array<string|Buffer>} ca
 * @see https.ServerOptions
 */
